src/erl/typespec.rs

```diff
@@ -0,0 +1,263 @@
+pub fn external_fun_spec(name: &str, args: &[ExternalFnArg], return_type: &Type) -> Document {
+    // Find all parameters used in the return type - these are the only ones we can show in the
+    // args otherwise they are flagged as unused
+    let return_type_params = extract_params_from_type(return_type);
+    "-spec "
+        .to_doc()
+        .append(atom(name.to_string()))
+        .append(external_spec_args(args, &return_type_params))
+        .append(" -> ")
+        .append(typ_to_erl(return_type))
+        .append(".")
+}
+
+fn external_spec_args(args: &[ExternalFnArg], allowed_params: &[String]) -> Document {
+    wrap_args(
+        args.iter()
+            .map(|arg| type_ast_name(&arg.typ, Some(allowed_params))),
+    )
+}
+
+// Extracts the type names from a TypeAst - We optionally pass in a list of type parameters that we
+// are aware of.
+fn type_ast_name(ast: &TypeAst, allowed_params: Option<&[String]>) -> Document {
+    match ast {
+        TypeAst::Var { name, .. } =>
+        // If we provide a list of allowed type parameter names then we're in a context with
+        // limited possible type parameter names and if we come across one that isn't 'allowed'
+        // then we're better just outputting 'any' instead. This prevents unknown or unused
+        // type parameter warnings from erlang
+        {
+            allowed_params
+                .map(|slice| {
+                    if slice.contains(name) {
+                        name.to_camel_case()
+                    } else {
+                        "any".to_string()
+                    }
+                })
+                .unwrap_or(name.to_camel_case())
+                .to_doc()
+        }
+        TypeAst::Constructor { name, args, .. } => type_to_erl(name, &args, allowed_params),
+        _ => "term".to_doc(),
+    }
+}
+
+fn type_to_erl(name: &str, args: &[TypeAst], allowed_params: Option<&[String]>) -> Document {
+    match name {
+        "String" => "binary()".to_doc(),
+        "Int" => "integer()".to_doc(),
+        "Bool" => "boolean()".to_doc(),
+        "List" => "list".to_doc().append(wrap_args(
+            args.iter().map(|arg| type_ast_name(arg, allowed_params)),
+        )),
+        "Map" => concat(
+            args.iter()
+                .map(|arg| type_ast_name(arg, allowed_params))
+                .intersperse(" => ".to_doc()),
+        )
+        .surround("#{", "}"),
+        _ => name.to_snake_case().to_doc(),
+    }
+}
+
+/* Extraction functions - There are various places where we want to make sure that we're only
+   outputting type parameters if we know they are going to be used in order to avoid unused type
+   parameter warnings. Consequently we have these functions to help us walk the type trees and find
+   what we need.
+*/
+fn extract_params_from_constructors(constructors: &[RecordConstructor]) -> Vec<String> {
+    let mut set = std::collections::HashSet::new();
```

Comment:
What is this set for? Should it be returned?

Reply:
It is a bit confusing and needs a comment but it was my attempt at a deduplicating the vector without sorting/reordering (which the standard `vec.dedup` method requires.) It is used in the `.filter(|name| set.insert(name.clone()))` below. It seems that `HashSet.insert` returns `true` if the item was not in the set and `false` if it was so I think it works as a filter predicate. There might well be better ways to approach it though.

Reply:
Hmm, thinking about this I suspect I'm trying to be careful with the wrong set of params. This 'collection' stage (if it is even needed) doesn't need to be order preserved. I'll try to explain my reasoning for all this with examples when I get a bit of time. It is either required due to certain complexities or possibly just a stop gap to cover an incomplete type conversion process.

Reply:
Actually for the use of it in `typespec::type_`, we want to be deduplicated and order preserved, I think. Might be better ways to handle it all though.